#!/usr/bin/env python3
"""
Executor ì‹¤í–‰ ê°€ì´ë“œ ë° í…ŒìŠ¤íŠ¸ ìŠ¤í¬ë¦½íŠ¸

ì´ ìŠ¤í¬ë¦½íŠ¸ëŠ” RL Arena Executorë¥¼ ë¡œì»¬ì—ì„œ ì‹¤í–‰í•˜ê³  í…ŒìŠ¤íŠ¸í•˜ëŠ” ë°©ë²•ì„ ë³´ì—¬ì¤ë‹ˆë‹¤.
"""

import subprocess
import sys
import time
import grpc
from pathlib import Path

# Proto ì»´íŒŒì¼ í™•ì¸
def check_proto_compiled():
    """Proto íŒŒì¼ì´ ì»´íŒŒì¼ë˜ì—ˆëŠ”ì§€ í™•ì¸"""
    executor_pb2 = Path("executor_pb2.py")
    executor_pb2_grpc = Path("executor_pb2_grpc.py")
    
    if not executor_pb2.exists() or not executor_pb2_grpc.exists():
        print("âŒ Proto íŒŒì¼ì´ ì»´íŒŒì¼ë˜ì§€ ì•Šì•˜ìŠµë‹ˆë‹¤.")
        print("ğŸ“ ë‹¤ìŒ ëª…ë ¹ì–´ë¥¼ ì‹¤í–‰í•˜ì„¸ìš”:")
        print("   make proto")
        print("   # ë˜ëŠ”")
        print("   python -m grpc_tools.protoc -I./proto --python_out=. --grpc_python_out=. --pyi_out=. ./proto/executor.proto")
        return False
    
    print("âœ… Proto íŒŒì¼ì´ ì»´íŒŒì¼ë˜ì–´ ìˆìŠµë‹ˆë‹¤.")
    return True


def check_dependencies():
    """í•„ìš”í•œ íŒ¨í‚¤ì§€ í™•ì¸"""
    try:
        import grpc
        import yaml
        print("âœ… í•„ìˆ˜ íŒ¨í‚¤ì§€ê°€ ì„¤ì¹˜ë˜ì–´ ìˆìŠµë‹ˆë‹¤.")
        return True
    except ImportError as e:
        print(f"âŒ í•„ìˆ˜ íŒ¨í‚¤ì§€ê°€ ì—†ìŠµë‹ˆë‹¤: {e}")
        print("ğŸ“ ë‹¤ìŒ ëª…ë ¹ì–´ë¥¼ ì‹¤í–‰í•˜ì„¸ìš”:")
        print("   pip install -r requirements.txt")
        return False


def setup_config_for_local():
    """ë¡œì»¬ ì‹¤í–‰ì„ ìœ„í•œ config ì„¤ì • í™•ì¸"""
    config_file = Path("config/limits.yaml")
    if not config_file.exists():
        print("âŒ config/limits.yaml íŒŒì¼ì´ ì—†ìŠµë‹ˆë‹¤.")
        return False
    
    print("âœ… Config íŒŒì¼ì´ ì¡´ì¬í•©ë‹ˆë‹¤.")
    print("ğŸ’¡ ë¡œì»¬ í…ŒìŠ¤íŠ¸ë¥¼ ìœ„í•´ config/limits.yamlì—ì„œ ë‹¤ìŒì„ í™•ì¸í•˜ì„¸ìš”:")
    print("   - executor.use_k8s: false  (K8s ì—†ì´ í…ŒìŠ¤íŠ¸)")
    print("   - sandbox.use_docker: false  (Docker ì—†ì´ í…ŒìŠ¤íŠ¸)")
    return True


def test_health_check(port=50051):
    """Executor ì„œë²„ Health Check"""
    try:
        import executor_pb2
        import executor_pb2_grpc
        
        channel = grpc.insecure_channel(f'localhost:{port}')
        stub = executor_pb2_grpc.ExecutorStub(channel)
        
        print(f"\nğŸ” Health Check í…ŒìŠ¤íŠ¸ (í¬íŠ¸ {port})...")
        request = executor_pb2.HealthCheckRequest()
        response = stub.HealthCheck(request, timeout=5)
        
        print(f"âœ… Health Check ì„±ê³µ!")
        print(f"   - Healthy: {response.healthy}")
        print(f"   - Version: {response.version}")
        print(f"   - Active Matches: {response.active_matches}")
        return True
        
    except grpc.RpcError as e:
        print(f"âŒ Health Check ì‹¤íŒ¨: {e.code()}")
        print(f"   {e.details()}")
        return False
    except Exception as e:
        print(f"âŒ ì˜¤ë¥˜ ë°œìƒ: {e}")
        return False


def print_usage():
    """ì‚¬ìš©ë²• ì¶œë ¥"""
    print("""
â•”â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•—
â•‘          RL Arena Executor ì‹¤í–‰ ê°€ì´ë“œ                     â•‘
â•šâ•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•

ğŸ“‹ ì‚¬ì „ ì¤€ë¹„ ë‹¨ê³„:

1ï¸âƒ£  Proto íŒŒì¼ ì»´íŒŒì¼
   make proto
   # ë˜ëŠ”
   python -m grpc_tools.protoc -I./proto --python_out=. --grpc_python_out=. --pyi_out=. ./proto/executor.proto

2ï¸âƒ£  Config ì„¤ì • (ë¡œì»¬ í…ŒìŠ¤íŠ¸ìš©)
   config/limits.yaml íŒŒì¼ì„ ìˆ˜ì •:
   
   # K8s ì—†ì´ í…ŒìŠ¤íŠ¸
   executor:
     use_k8s: false
   
   # Docker ì—†ì´ í…ŒìŠ¤íŠ¸ (ì„ íƒì‚¬í•­)
   sandbox:
     use_docker: false

3ï¸âƒ£  ì˜ì¡´ì„± ì„¤ì¹˜ í™•ì¸
   pip install -r requirements.txt

ğŸš€ Executor ì‹¤í–‰ ë°©ë²•:

ë°©ë²• 1: ì§ì ‘ ì‹¤í–‰
   python -m executor.server
   # ë˜ëŠ” í™˜ê²½ë³€ìˆ˜ì™€ í•¨ê»˜
   EXECUTOR_HOST=0.0.0.0 EXECUTOR_PORT=50051 LOG_LEVEL=INFO python -m executor.server

ë°©ë²• 2: Make ì‚¬ìš© (Makefileì´ ìˆëŠ” ê²½ìš°)
   make run

ë°©ë²• 3: ë°±ê·¸ë¼ìš´ë“œ ì‹¤í–‰
   nohup python -m executor.server > executor.log 2>&1 &

ğŸ§ª ì‹¤í–‰ í™•ì¸:

ë‹¤ë¥¸ í„°ë¯¸ë„ì—ì„œ:
   python examples/test_executor.py
   # ë˜ëŠ”
   grpcurl -plaintext localhost:50051 list

ğŸ“ ì£¼ìš” í™˜ê²½ë³€ìˆ˜:
   EXECUTOR_HOST=0.0.0.0      # ì„œë²„ í˜¸ìŠ¤íŠ¸
   EXECUTOR_PORT=50051         # gRPC í¬íŠ¸
   CONFIG_PATH=config/limits.yaml  # Config íŒŒì¼ ê²½ë¡œ
   LOG_LEVEL=INFO              # ë¡œê·¸ ë ˆë²¨

ğŸ”§ íŠ¸ëŸ¬ë¸”ìŠˆíŒ…:

1. "Proto íŒŒì¼ ì—†ìŒ" ì—ëŸ¬
   â†’ make proto ì‹¤í–‰

2. "ModuleNotFoundError: No module named 'executor_pb2'"
   â†’ make proto ì‹¤í–‰ í›„ ë‹¤ì‹œ ì‹œë„

3. "Address already in use" ì—ëŸ¬
   â†’ í¬íŠ¸ê°€ ì‚¬ìš© ì¤‘ì…ë‹ˆë‹¤. ë‹¤ë¥¸ í¬íŠ¸ ì‚¬ìš©:
     EXECUTOR_PORT=50052 python -m executor.server

4. "Kubernetes connection failed"
   â†’ config/limits.yamlì—ì„œ use_k8s: false ì„¤ì •

5. "Docker connection failed"
   â†’ config/limits.yamlì—ì„œ use_docker: false ì„¤ì •
   â†’ ë˜ëŠ” Docker ë°ëª¬ ì‹¤í–‰

ğŸ“š ë” ë§ì€ ì •ë³´:
   - README.md
   - docs/K8S_ARCHITECTURE.md
   - docs/K8S_SETUP.md
    """)


def main():
    """ë©”ì¸ ì‹¤í–‰"""
    print("ğŸ” RL Arena Executor ì‹¤í–‰ ì „ ê²€ì‚¬...\n")
    
    # 1. Proto ì»´íŒŒì¼ í™•ì¸
    proto_ok = check_proto_compiled()
    
    # 2. ì˜ì¡´ì„± í™•ì¸
    deps_ok = check_dependencies()
    
    # 3. Config í™•ì¸
    config_ok = setup_config_for_local()
    
    print("\n" + "="*60)
    
    if not proto_ok:
        print("\nâŒ Proto íŒŒì¼ì„ ë¨¼ì € ì»´íŒŒì¼í•´ì•¼ í•©ë‹ˆë‹¤.")
        print("   ì‹¤í–‰: make proto")
        sys.exit(1)
    
    if not deps_ok:
        print("\nâŒ í•„ìˆ˜ íŒ¨í‚¤ì§€ë¥¼ ë¨¼ì € ì„¤ì¹˜í•´ì•¼ í•©ë‹ˆë‹¤.")
        print("   ì‹¤í–‰: pip install -r requirements.txt")
        sys.exit(1)
    
    if not config_ok:
        print("\nâŒ Config íŒŒì¼ì„ í™•ì¸í•´ì£¼ì„¸ìš”.")
        sys.exit(1)
    
    print("\nâœ… ëª¨ë“  ì‚¬ì „ ì¤€ë¹„ê°€ ì™„ë£Œë˜ì—ˆìŠµë‹ˆë‹¤!")
    print("\n" + "="*60)
    print_usage()
    
    # ì„œë²„ê°€ ì‹¤í–‰ ì¤‘ì¸ì§€ í™•ì¸
    print("\nğŸ” ì„œë²„ ì‹¤í–‰ ì—¬ë¶€ í™•ì¸...")
    if test_health_check():
        print("\nâœ… Executorê°€ ì´ë¯¸ ì‹¤í–‰ ì¤‘ì…ë‹ˆë‹¤!")
    else:
        print("\nğŸ’¡ Executorë¥¼ ì‹¤í–‰í•˜ë ¤ë©´:")
        print("   python -m executor.server")


if __name__ == "__main__":
    main()
